### ISSUE TITLE
Calico pods stuck in CrashLoopBackOff in AKS 1.20.7

### ISSUE THREAD

**Issue by guy-microsoft** [Azure/AKS#2450]
**What happened**:
Upgraded an AKS cluster from 1.18.X to 1.19.11 and then to 1.20.7. After upgrading to 1.20.7 I see that some of the Calico pods are stuck in an infinite CrashLoopBackOff.
In their container logs I see (looks like something with IP collision between 2 nodes):
2021-07-18 06:11:58.517 [INFO][8] startup/startup.go 390: Early log level set to info
2021-07-18 06:11:58.517 [INFO][8] startup/startup.go 406: Using NODENAME environment for node name mynodep-75352824-vmss000001
2021-07-18 06:11:58.517 [INFO][8] startup/startup.go 418: Determined node name: mynodep-75352824-vmss000001
2021-07-18 06:11:58.517 [INFO][8] startup/startup.go 103: Starting node mynodep-75352824-vmss000001 with version v3.19.0-dirty
2021-07-18 06:11:58.519 [INFO][8] startup/startup.go 450: Checking datastore connection
2021-07-18 06:11:58.576 [INFO][8] startup/startup.go 474: Datastore connection verified
2021-07-18 06:11:58.576 [INFO][8] startup/startup.go 113: Datastore is ready
2021-07-18 06:11:58.622 [INFO][8] startup/startup.go 515: Initialize BGP data
2021-07-18 06:11:58.622 [INFO][8] startup/startup.go 774: Using autodetected IPv4 address on interface eth0: 10.240.0.5/16
2021-07-18 06:11:58.622 [INFO][8] startup/startup.go 591: Node IPv4 changed, will check for conflicts
2021-07-18 06:11:58.644 [WARNING][8] startup/startup.go 1135: Calico node 'mynodep-75352824-vmss000000' is already using the IPv4 address 10.240.0.5.
2021-07-18 06:11:58.644 [INFO][8] startup/startup.go 354: Clearing out-of-date IPv4 address from this node IP="10.240.0.5/16"
2021-07-18 06:11:58.680 [WARNING][8] startup/startup.go 1347: Terminating
Calico node failed to start

**What you expected to happen**:
I expect that after upgrading AKS to 1.20.7 Calico pods will work as before.

**How to reproduce it (as minimally and precisely as possible)**:
Create a cluster with K8S 1.18 (more details below) and then upgrade to 1.19.11 and then to 1.20.7. Check for Calico pods in the calico-system namespace.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.20.7
- Size of cluster (how many worker nodes are in the cluster?): 2 node pools with auto scaling, one "System" and one "User" with Min count 3 max count 100 for both.
- Others: Location: EastUS

**Comment by justindavies** [comment]
Would you be able to confirm if there was an IP collision in the VMSS as you thought?  This will help us get to the bottom of this

**Comment by guy-microsoft** [comment]
@justindavies It seems that the nodes InternalIPs are different:
mynodep-75352824-vmss000001 InternalIP: 10.240.0.5
mynodep-75352824-vmss000000 InternalIP: 10.240.0.4

However, in mynodep-75352824-vmss000000 I see the following annotation:
`projectcalico.org/IPv4Address: 10.240.0.5/16`
which seems to point to the InternalIP of mynodep-75352824-vmss000001.
I'm not familiar with what this annotation means, but I have a feeling that it's related to the issue.

**Comment by scoof** [comment]
We also hit this, and have an open support case regarding it, where we provided kubelet logs. There were no IP collisions, but the node ips in the Calico database were wrong. Restarting the calico pods on the nodes with the wrong recorded IP (i.e. not the host in CrashLoopBackoff) resolved it. Support request id: 121071925000234

**Comment by MarcosMMartinez** [comment]
Hit on 2107230040003040.

**Comment by asubmani** [comment]
> We also hit this, and have an open support case regarding it, where we provided kubelet logs. There were no IP collisions, but the node ips in the Calico database were wrong. Restarting the calico pods on the nodes with the wrong recorded IP (i.e. not the host in CrashLoopBackoff) resolved it. Support request id: 121071925000234
> We hit this issue when upgrading from 1.20.5 to 1.21.1, and respinning the 1.20.5 cluster and doing the upgrade again did not trigger the issue, so it seems to be timing-related.

Did you have to re-create the cluster?

**Comment by scoof** [comment]
> > We also hit this, and have an open support case regarding it, where we provided kubelet logs. There were no IP collisions, but the node ips in the Calico database were wrong. Restarting the calico pods on the nodes with the wrong recorded IP (i.e. not the host in CrashLoopBackoff) resolved it. Support request id: 121071925000234
> > We hit this issue when upgrading from 1.20.5 to 1.21.1, and respinning the 1.20.5 cluster and doing the upgrade again did not trigger the issue, so it seems to be timing-related.
> 
> Did you have to re-create the cluster?

To fix it? No, just restart the calico nodes that are registered with a wrong IP

**Comment by ekarlso** [comment]
We've had this issue in several clusters after going to 1.20.x with it seemingly looking ok and then starting to crash randomly

**Comment by JJ11teen** [comment]
We're seeing the same issue, and can confirm that restarting the non-failing calico pods seems to fix it.

I've had it crop up twice with the same cluster when scaling from 2->3 nodes, but I only restarted 1 (working) set of calico pods each time, so its possible that it just clashed with each IP in turn. Will update if it crops up again now all calico pods have been restarted.

**Comment by guy-microsoft** [comment]
@justindavies I found this comment in a relevant Calico github issue:
https://github.com/projectcalico/calico/issues/4611#issuecomment-880018862
Does it mean I should expect issue to be resolved?

**Comment by palma21** [comment]
Yes, this should be fixed now for all new upgrades, anyone hitting it from past upgrades can still restart to fix it.

**Comment by arbhatta-blk** [comment]
Ran into the same issue as described above. Restarting the pods solved the issue as suggested above.

Running the following command restarted the pods and fixed the issue for me:
`kubectl [-n K8s_Namespace_where_pods_are_failing] delete pods -l "k8s-app=calico-node"`

_Prior to restarting, I also observed a difference in `pod-template-generation` value in the pod that was in `CrashLoopBackOff` state_

Running the following command prior to restarting showed that the Running pods had a `pod-template-generation=1` label where-as the pod in `CrashLoopBackOff` had `pod-template-generation=5` 
`kubectl [-n K8s_Namespace_where_pods_are_failing] get pods -l "k8s-app=calico-node" --show-labels`

After the restart, all the pods had status RUNNING and all of them had the label `pod-template-generation=5`.
